cups/ppd-cache.c

```diff
@@ -1967,14 +2015,988 @@ _ppdCacheCreateWithPPD(ppd_file_t *ppd)	/* I - PPD file */
+	  default_ch = k;
+	}
+
+       /*
+	* Color/Gray - print-color-mode
+	*/
+
+	/* If we have a color device, check whether this option sets mono or
+	   color printing */
+	if (is_color)
+	{
+	  if (_cups_strcasecmp(o, "CNIJSGrayScale") == 0)
+	  {
+	    if (_cups_strcasecmp(c, "1") == 0)
+	      properties->sets_mono = 2;
+	    else
+	      properties->sets_color = 1;
+	  }
+	  else if (_cups_strcasecmp(o, "HPColorAsGray") == 0) /* HP PostScript */
+	  {
+	    if (_cups_strcasecmp(c, "True") == 0)
+	      properties->sets_mono = 2;
+	    else
+	      properties->sets_color = 1;
+	  }
+	  else if (_cups_strcasecmp(o, "ColorModel") == 0 ||
+		   _cups_strcasecmp(o, "ColorMode") == 0 ||
+		   _cups_strcasecmp(o, "OutputMode") == 0 ||
+		   _cups_strcasecmp(o, "PrintoutMode") == 0 ||
+		   _cups_strcasecmp(o, "ARCMode") == 0 || /* Sharp */
+		   strcasestr(o, "ColorMode") ||
+		   _cups_strcasecmp(o, "ColorResType") == 0 || /* Toshiba */
+		   strcasestr(o, "MonoColor")) /* Brother */
+	  {
+	    /* Monochrome/grayscale printing */
+	    if (strcasestr(c, "Mono") ||
+		_cups_strcasecmp(c, "Black") == 0 ||
+		((p = strcasestr(c, "Black")) && strcasestr(p, "White")) ||
+		(_cups_strncasecmp(c, "BW", 2) == 0 && !isalpha(c[2])))
+	      properties->sets_mono = 2;
+	    else if (strcasestr(c, "Gray") ||
+		     strcasestr(c, "Grey") ||
+		     _cups_strcasecmp(c, "BlackOnly") == 0) /* Lexmark */
+	      properties->sets_mono = 3;
+
+	    /* Color printing */
+	    if (((p = strcasestr(c, "CMY")) && !strcasestr(p, "Gray")) ||
+		_cups_strcasecmp(c, "ColorOnly") == 0 || /* Lexmark */
+		((p = strcasestr(c, "Adobe")) && strcasestr(p, "RGB")))
+	      properties->sets_color = 2;
+	    else if (strcasestr(c, "sRGB"))
+	      properties->sets_color = 4;
+	    else if (strcasestr(c, "RGB") ||
+		     strcasestr(c, "Color"))
+	      properties->sets_color = 3;
+	  }
+
+	  /* This option actually sets color mode */
+	  if (properties->sets_mono || properties->sets_color)
+	    sets_color_mode = 1;
+	}
+
+       /*
+	* Output Quality - print-quality
+	*/
+
+	/* check whether this option affects print quality or content
+	   optimization */
+
+	/* Determine influence of the options and choices on the print
+	   quality by their names */
+
+	/* Vendor-specific option and choice names */
+	if (_cups_strcasecmp(o, "ARCPPriority") == 0) /* Sharp */
+	{
+	  if (_cups_strcasecmp(c, "Quality") == 0)
+	    properties->sets_high = 10;
+	  else if (_cups_strcasecmp(c, "Speed") == 0)
+	    properties->sets_draft = 10;
+	}
+	else if (_cups_strcasecmp(o, "BRJpeg") == 0) /* Brother */
+	{
+	  if (_cups_strcasecmp(c, "QualityPrior") == 0)
+	    properties->sets_high = 10;
+	  else if (_cups_strcasecmp(c, "SpeedPrior") == 0)
+	    properties->sets_draft = 10;
+	}
+	else if (_cups_strcasecmp(o, "RIPrintMode") == 0) /* Ricoh & OEM */
+	{
+	  if (_cups_strcasecmp(c, "1rhit") == 0)
+	    properties->sets_high = 7;
+	  else if (_cups_strcasecmp(c, "6rhit") == 0)
```

Comment:
Don't try to do so much with presets. Keep them simple, otherwise there are just too many presets!

Reply:
The lots of rules which some very specific ones come from my testing on existing PPD files.  The one you cite is for Ricoh, and they put out very many PPDs for their professional printers, several PPDs (for the different PDLs) for each new model and also fixes for existing PPDs).

I tested several for PostScript (HP, Ricoh, Toshiba, ...), HPLIP, Gutenprint, Foomatic, ... and this made me adding some "quirk" rules to the originally written up ones.

Do not think that in the future we have to update these rules all the time for the new PPDs. As PPDs are deprecated manufacturers an driver developers will hopefully not continue developing on them and squeeze in their required functionality with more and more adventurous workarounds. The rules are a good set for handling the legacy of PPDs and giving the user a modern IPP printing experience.

Reply:
Till, presets are most useful for common combinations of options. Complex printers could easily provide thousands of potential "presets" with your current code, and not all of those combinations are valid (need to check for constraints, which on these devices are also complex and heavily optimized by me years ago).

I agree that we should do everything we can to offer consistency to users for all printers, but *that* functionality/consistency is targeted at general users and not the "experts" that are using those high-end printers.

Reply:
For the experts I always leave the possibility to set the options manually. If an option is supplied with the job, it will not get auto-set by the pre-set.
Do you have some sample PPDs where there are complex constraints which could interfere with auto-created pre-sets? I have only seen constraints with installable accessories, or constraints with paper source/paper type, paper size/duplex, like duplex on transparencies or A3 paper from envelope tray, not on resolution,. toner saving, resolution enhancement, ...